frontend/ui/widget/virtualkeyboard.lua

```diff
@@ -1035,6 +1040,10 @@ function VirtualKeyboard:goToStartOfLine()
     self.inputbox:goToStartOfLine()
 end
 
+function VirtualKeyboard:utf8modeChar()
+    if self.inputbox.utf8modeChar then self.inputbox:utf8modeChar() end
```

Comment:
A bit confused what this does. A verbal method name could have make it clearer. emitUtf8ModeKey ?
Anyway, a small comment feels needed :)

Reply:
I was trying to clean the stage (remove candidates) when tapping on the Earth button. A verbal one is indeed better.

Reply:
So, may be it should be named as what it does ? `:cleanStage()` (or more explicite, I dunno what "stage" means :).
And may be in the zh virtualkeyboard, the utf8key should call cleanStage() ?

Or does calling emitUtf8Key() really means something for other keyboards ?

Reply:
In zh virtualkeyboard, the `util.wrapMethod` setup is uft8key calls `wrappedSeparate` (which does the same thing as typing Separation thus cleanStage) before calling the actual `emitUtf8ModeKey`.

With show candidates on, you type stroke `一` and get character `一` in the unfinished state (so you can add strokes to it or finish it to start a new character.) Then type `丨` and get `十[丁丅]` which says the current on-stage character is `十` but there are two other candidates "off-stage". The extra `[丁丅]` is not part of the actual text so will need to be cleaned when a finishing key or other none-stroke key is hit, which includes the uft8 key. Otherwise the off-stage characters are left on-stage to mess with user operations.

Reply:
By the way, the in-place candidates is my design :) In other similar cases, there would be a section on top of the keys dedicated to showing candidates so there's no need to clean the actual text.

Reply:
> By the way, the in-place candidates is my design :) 

I find it quite nice: you don't have to move your eyes away from the line you are typing.

I dunno how this feels to native Chinese writers, but when testing, and in `十[丁丅]`, it was a bit hard for me to distinguish the brackets from these Chinese quite straight glyphs. (Dunno if other unicode chars with more spacing or more or less thickness could help.)

Reply:
I tried different brackets, the current [] looks best. It's easy to distinguish because they are out of the grid :)